regulation-worker/internal/delete/batch/batch.go

```diff
@@ -0,0 +1,379 @@
+//returns list of all .json.gz files and marks exists as true if `statusTrackerFile` is present in the destination.
+//NOTE: assuming that all of batch destination have same file system as S3, i.e. flat.
+func (b *Batch) listFilesAndCheckTrackerFile(ctx context.Context) ([]*filemanager.FileObject, bool, error) {
+
+	fileObjects, err := b.FM.ListFilesWithPrefix("", listMaxItem)
+	if err != nil {
+		return []*filemanager.FileObject{}, false, fmt.Errorf("failed to fetch object list from S3:%w", err)
+	}
+
+	//since everything is stored as a file in S3, above fileObjects list also has directory & not just *.json.gz files. So, need to remove those.
+	count := 0
+	exist := false
+	regexRequiredSuffix := regexp.MustCompile(".json.gz$")
+	for i := 0; i < len(fileObjects); i++ {
+		if regexRequiredSuffix.Match([]byte(fileObjects[i].Key)) {
+			count++
+		}
+		if fileObjects[i].Key == statusTrackerFile {
+			exist = true
+		}
+	}
+	//list of only .gz files
+	gzFileObjects := make([]*filemanager.FileObject, count)
+	index := 0
+	for i := 0; i < len(fileObjects); i++ {
+		if regexRequiredSuffix.Match([]byte(fileObjects[i].Key)) {
+			gzFileObjects[index] = fileObjects[i]
+			index++
+		}
+	}
+	return gzFileObjects, exist, nil
+}
+
+//two pointer algorithm implementation to remove all the files from which users are already deleted.
+func removeCleanedFiles(files []*filemanager.FileObject, cleanedFiles []string) []*filemanager.FileObject {
+	sort.Slice(files, func(i, j int) bool {
+		return files[i].Key < files[j].Key
+	})
+	sort.Slice(cleanedFiles, func(i, j int) bool {
+		return cleanedFiles[i] < cleanedFiles[j]
+	})
+
+	i := 0
+	j := 0
+	presentCount := 0
+	present := make([]bool, len(files))
+	for j < len(cleanedFiles) {
+		if files[i].Key < cleanedFiles[j] {
+			i++
+		} else if files[i].Key > cleanedFiles[j] {
+			j++
+		} else {
+			present[i] = true
+			presentCount++
+			i++
+			j++
+		}
+	}
+	j = 0
+	finalFiles := make([]*filemanager.FileObject, len(files)-presentCount)
+
+	for i := 0; i < len(files); i++ {
+		if !present[i] {
+			finalFiles[j] = files[i]
+			j++
+		}
+	}
+	return finalFiles
+}
+
+//append <fileName> to <statusTrackerFile> locally for which deletion has completed.
+func updateStatusTrackerFile(fileName string) error {
+	f, err := os.OpenFile(statusTrackerFile, os.O_APPEND|os.O_WRONLY, 0644)
+	if err != nil {
+		return fmt.Errorf("error while opening statusTrackerFile: %w", err)
+	}
+	if _, err := f.Write([]byte(fileName)); err != nil {
+		return fmt.Errorf("error while writing to statusTrackerFile: %w", err)
+	}
+	if _, err := f.Write([]byte("\n")); err != nil {
+		return fmt.Errorf("error while writing to statusTrackerFile: %w", err)
+	}
+	if err := f.Close(); err != nil {
+		return fmt.Errorf("error while closing statusTrackerFile: %w", err)
+	}
+	return nil
+}
+
+//downloads `fileName` locally
+//Note: download happens concurrently in 5 go routine by default
+func (b *Batch) download(ctx context.Context, fileName string) error {
```

Comment:
I believe it would be better if download, was using a specific directory to store all the files, but default a random folder insider tmp.

Moreover use random file names for the downloaded files to avoid naming conflicts.

the `Batch` struct will have an optional string to store the directory and download will return the full path of the random file name.

Is important for those files to get deleted in case of a crash for privacy and regulatory reasons. In most cases tmp directory gets cleared after restart so it is safer there.